File: src/socket_example.py
#!/usr/bin/env python3
'''
A stripped down version of ardupilot/libraries/SITL/examples/JSON/pybullet/robot.py

There is no physics engine or robot model - just the SITL-JSON socket interface

Usage
-----

1. Run the example:

    $ socket_example.py  

2. Start SITL:

    $ sim_vehicle.py -v Rover -f JSON:127.0.0.1 --aircraft=sitl_json

'''

import socket
import struct
import json

# use pymavlink for ArduPilot convention transformations
from pymavlink.rotmat import Vector3, Matrix3
from pymavlink.quaternion import Quaternion

import time

RATE_HZ = 50.0
TIME_STEP = 1.0 / RATE_HZ
GRAVITY_MSS = 9.80665

time_now = 0
last_velocity = None

def to_tuple(vec3):
    '''convert a Vector3 to a tuple'''
    return (vec3.x, vec3.y, vec3.z)

def init():
  global time_now
  time_now = 0

def physics_step(pwm_in):

  global time_now
  time_now += TIME_STEP

  quaternion = Quaternion([1, 0, 0, 0])
  roll, pitch, yaw = quaternion.euler
  velocity = Vector3(0, 0, 0)
  position = Vector3(0, 0, 0)
  gyro = Vector3(0, 0, 0)

  # calculate acceleration
  global last_velocity
  if last_velocity is None:
      last_velocity = velocity

  accel = (velocity - last_velocity) * (1.0 / TIME_STEP)
  last_velocity = velocity

  # add in gravity in earth frame
  accel.z -= GRAVITY_MSS

  # convert accel to body frame
  # accel = dcm.transposed() * accel

  # convert to tuples
  accel = to_tuple(accel)
  gyro = to_tuple(gyro)
  position = to_tuple(position)
  velocity = to_tuple(velocity)
  euler = (roll, pitch, yaw)

  return time_now,gyro,accel,position,euler,velocity

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind(('', 9002))
sock.settimeout(0.1)

last_SITL_frame = -1
connected = False
frame_count = 0
frame_time = time.time()
print_frame_count = 1000

while True:

  py_time = time.time()

  try:
      data,address = sock.recvfrom(100)
  except Exception as ex:
      time.sleep(0.01)
      continue

  parse_format = 'HHI16H'
  magic = 18458

  if len(data) != struct.calcsize(parse_format):
    print("got packet of len %u, expected %u" % (len(data), struct.calcsize(parse_format)))
    continue

  decoded = struct.unpack(parse_format,data)

  if magic != decoded[0]:
      print("Incorrect protocol magic %u should be %u" % (decoded[0], magic))
      continue

  frame_rate_hz = decoded[1]
  frame_count = decoded[2]
  pwm = decoded[3:]

  if frame_rate_hz != RATE_HZ:
      print("New frame rate %u" % frame_rate_hz)
      RATE_HZ = frame_rate_hz
      TIME_STEP = 1.0 / RATE_HZ

  # Check if the fame is in expected order
  if frame_count < last_SITL_frame:
    # Controller has reset, reset physics also
    init()
    print('Controller reset')
  elif frame_count == last_SITL_frame:
    # duplicate frame, skip
    print('Duplicate input frame')
    continue
  elif frame_count != last_SITL_frame + 1 and connected:
    print('Missed {0} input frames'.format(frame_count - last_SITL_frame - 1))
  last_SITL_frame = frame_count

  if not connected:
    connected = True
    print('Connected to {0}'.format(str(address)))
  frame_count += 1

  # physics time step
  phys_time,gyro,accel,pos,euler,velo = physics_step(pwm)

  # build JSON format
  IMU_fmt = {
    "gyro" : gyro,
    "accel_body" : accel
  }
  JSON_fmt = {
    "timestamp" : phys_time,
    "imu" : IMU_fmt,
    "position" : pos,
    "attitude" : euler,
    "velocity" : velo
  }
  JSON_string = "\n" + json.dumps(JSON_fmt,separators=(',', ':')) + "\n"

  # Send to AP
  sock.sendto(bytes(JSON_string,"ascii"), address)

  # Track frame rate
  if frame_count % print_frame_count == 0:
    now = time.time()
    total_time = now - frame_time
    print("%.2f fps T=%.3f dt=%.3f" % (print_frame_count/total_time, phys_time, total_time))
    frame_time = now

File: src/SocketExample.cc
// Test socket connection to SITL
//
// Use a copy of the ArduPilot SocketAPM class from
// ardupilot/libraries/AP_HAL/utility/Socket.h
//
// This example is based on socket_example.py which in turn is a
// simplified copy of ardupilot/libraries/SITL/examples/JSON/pybullet/robot.py
//
// Usage
// -----
// 
// 1. Build and run the example:
//
//  $ mkdir build && cd build
//  $ cmake ..
//  $ make
//  $ ./SITLSocketExample  
// 
// 2. Start SITL:
//
//  $ sim_vehicle.py -v Rover -f JSON:127.0.0.1 --aircraft=sitl_json
//
#include <Socket.h>

#include <chrono>
#include <cinttypes>
#include <exception>
#include <iostream>
#include <string>
#include <thread>

#include <rapidjson/stringbuffer.h>
#include <rapidjson/writer.h>

const double GRAVITY_MSS = 9.80665;

// SIM_JSON.h
struct servo_packet {
    uint16_t magic;         // 18458 expected magic value
    uint16_t frame_rate;
    uint32_t frame_count;
    uint16_t pwm[16];
};

int main(int argc, const char* argv[])
{
    try
    {
        // configure socket 
        std::string address = "127.0.0.1";
        uint16_t port = 9002;
        SocketAPM sock(true);
        // sock.set_blocking(false);
        // sock.reuseaddress();
        sock.bind(address.c_str(), port);
        std::cout << "SITL interface set to " << address << ":" << port << "\n"; 

        // sitl input
        bool sitl_connected = false;
        uint16_t sitl_frame_rate = 50;
        uint32_t sitl_frame_count = 0;
        uint32_t sitl_last_frame = -1;
        uint16_t sitl_pwm[16];

        // start clocks
        uint16_t update_rate_hz = 50;
        auto update_duration = std::chrono::milliseconds(1000/update_rate_hz);
        auto sim_start_time = std::chrono::steady_clock::now();        
        auto last_update_time = sim_start_time; 

        while (true)
        {
            auto now = std::chrono::steady_clock::now();
            auto sim_time = now - sim_start_time;
            auto sim_time_ns = std::chrono::duration_cast<std::chrono::nanoseconds>(sim_time);
            // std::cout << "sim_time_ns: " << sim_time_ns.count() << "\n";

            servo_packet pkt;
            ssize_t bytes_recv = 0;
            const char* ip;
            try
            {   
                bytes_recv = sock.recv(&pkt, sizeof(pkt), 100);
                sock.last_recv_address(ip, port);
            }
            catch(...)
            {
                using namespace std::chrono_literals;
                std::this_thread::sleep_for(1000ms);
                continue;
            }

            if (bytes_recv >= 0)
            {
                // inspect sitl packet
                std::cout << "recv " << bytes_recv << " bytes from " << ip << ":" << port << "\n";
                std::cout << "magic: " << pkt.magic << "\n";
                std::cout << "frame_rate: " << pkt.frame_rate << "\n";
                std::cout << "frame_count: " << pkt.frame_count << "\n";
                std::cout << "pwm: [";
                for (auto i=0; i<15; ++i)
                {
                    std::cout << pkt.pwm[i] << ", ";
                }
                std::cout << pkt.pwm[15] << "]\n";

                // check magic
                uint16_t magic = 18458;
                if (magic != pkt.magic)
                {
                    std::cout << "Incorrect protocol magic "
                        << pkt.magic << " should be "
                        << magic << "\n";
                    continue; 
                }

                // capture sitl packet
                sitl_frame_rate = pkt.frame_rate;
                sitl_frame_count = pkt.frame_count;
                std::copy(std::begin(pkt.pwm), std::end(pkt.pwm), std::begin(sitl_pwm));

                // check frame rate
                if (pkt.frame_rate != sitl_frame_rate)
                {
                    std::cout << "New frame rate " << pkt.frame_rate << "\n";
                    sitl_frame_rate = pkt.frame_rate;
                }

                // check frame order
                if (pkt.frame_count < sitl_last_frame)
                {
                    // controller has reset
                    std::cout << "Controller reset\n";
                }
                else if (pkt.frame_count == sitl_last_frame)
                {
                    // recv duplicate frame, skip
                    std::cout << "Duplicate input frame\n";
                    // continue;
                }
                else if (pkt.frame_count != sitl_last_frame + 1 && sitl_connected)
                {
                    // missed frame
                    std::cout << "Missed "
                        << pkt.frame_count - sitl_last_frame - 1
                        << " input frames\n";
                }
                sitl_last_frame = pkt.frame_count;

                if (!sitl_connected)
                {
                    sitl_connected = true;
                    std::cout << "Connected to " << address << "\n";
                }

                // update and send JSON
                // if (now - last_update_time > update_duration)
                {    
                    using namespace rapidjson;

                    last_update_time += update_duration;

                    // update physics for time step here

                    // require the duration since sim start in seconds 
                    double timestamp = sim_time_ns.count() * 1.0E-9;

                    // build JSON document
                    StringBuffer s;
                    Writer<StringBuffer> writer(s);            

                    writer.StartObject();

                    writer.Key("timestamp");
                    writer.Double(timestamp);

                    writer.Key("imu");
                    writer.StartObject();
                    writer.Key("gyro");
                    writer.StartArray();
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.EndArray();
                    writer.Key("accel_body");
                    writer.StartArray();
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.Double(-GRAVITY_MSS);
                    writer.EndArray();
                    writer.EndObject();

                    writer.Key("position");
                    writer.StartArray();
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.EndArray();

                    // ArduPilot quaternion convention: q[0] = 1 for identity.
                    writer.Key("quaternion");
                    writer.StartArray();
                    writer.Double(1.0);
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.EndArray();

                    writer.Key("velocity");
                    writer.StartArray();
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.Double(0.0);
                    writer.EndArray();

                    writer.Key("rng_1");
                    writer.Double(0.0);

                    writer.Key("rng_1");
                    writer.Double(0.0);

                    writer.Key("windvane");
                    writer.StartObject();
                    writer.Key("direction");
                    writer.Double(1.57079633);
                    writer.Key("speed");
                    writer.Double(5.5);
                    writer.EndObject();

                    // send JSON
                    std::string json_str = "\n" + std::string(s.GetString()) + "\n";
                    auto bytes_sent = sock.sendto(json_str.c_str(), json_str.size(), address.c_str(), port);

                    std::cout << "sent " << bytes_sent <<  " bytes to " 
                        << ip << ":" << port << "\n";
                    std::cout << json_str << "\n";

                }
            }
        }
    }
    catch(const std::exception& e)
    {
        std::cerr << e.what() << '\n';
    }

    return 0;
}
